File: Sources/CuotaX/Quota.swift
// SPDX-License-Identifier: MIT

import Foundation

let fiveHourWindowDurationMinutes = 5 * 60
let weeklyWindowDurationMinutes = 7 * 24 * 60

struct QuotaWindow: Equatable, Sendable {
  let usedPercent: Double
  let resetsAt: Date?
}

struct Quota: Equatable, Sendable {
  let fiveHour: QuotaWindow?
  let weekly: QuotaWindow?
  let updatedAt: Date

  var highestPercent: Double? {
    [fiveHour, weekly].compactMap(\.self).map(\.usedPercent).max()
  }
}

enum QuotaLevel: Sendable {
  case normal
  case warning
  case critical
  case unavailable
}

struct QuotaStatus: Sendable {
  let coverage: Double?
  let level: QuotaLevel
  let onTrackPercent: Double?
  let remainingPercent: Double?
  let window: String?

  var paceLabel: String {
    switch level {
    case .normal: "On track"
    case .warning: "Running high"
    case .critical: "At risk"
    case .unavailable: "Pace unavailable"
    }
  }
}

extension Quota {
  func status(now: Date = Date()) -> QuotaStatus {
    let statuses = [
      windowStatus(
        fiveHour, name: "5-hour", durationMinutes: fiveHourWindowDurationMinutes, now: now),
      windowStatus(
        weekly, name: "weekly", durationMinutes: weeklyWindowDurationMinutes, now: now),
    ].compactMap(\.self)

    guard !statuses.isEmpty else {
      return QuotaStatus(
        coverage: nil,
        level: .unavailable,
        onTrackPercent: nil,
        remainingPercent: nil,
        window: nil
      )
    }

    let timed = statuses.filter { $0.onTrackPercent != nil }
    return (timed.isEmpty ? statuses : timed).min { $0.coverage < $1.coverage }!.status
  }
}

private struct RankedStatus {
  let coverage: Double
  let status: QuotaStatus

  var onTrackPercent: Double? { status.onTrackPercent }
}

private func windowStatus(
  _ value: QuotaWindow?,
  name: String,
  durationMinutes: Int,
  now: Date
) -> RankedStatus? {
  guard let value else { return nil }

  let usedPercent = normalizedPercent(value.usedPercent)
  let remainingPercent = 100 - usedPercent
  guard let resetsAt = value.resetsAt, resetsAt > now else {
    return RankedStatus(
      coverage: remainingPercent / 100,
      status: QuotaStatus(
        coverage: remainingPercent / 100,
        level: severity(usedPercent),
        onTrackPercent: nil,
        remainingPercent: remainingPercent,
        window: name
      )
    )
  }

  let duration = Double(durationMinutes * 60)
  let timeRemaining = min(1, max(0, resetsAt.timeIntervalSince(now) / duration))
  let onTrackPercent = (1000 * (1 - timeRemaining)).rounded() / 10
  let coverage = remainingPercent / 100 / timeRemaining
  let level: QuotaLevel =
    if usedPercent <= onTrackPercent {
      .normal
    } else if coverage < 0.5 {
      .critical
    } else {
      .warning
    }

  return RankedStatus(
    coverage: coverage,
    status: QuotaStatus(
      coverage: coverage,
      level: level,
      onTrackPercent: onTrackPercent,
      remainingPercent: remainingPercent,
      window: name
    )
  )
}

struct PaceColor: Equatable, Sendable {
  let red: Int
  let green: Int
  let blue: Int
}

func paceColor(_ status: QuotaStatus) -> PaceColor? {
  guard status.level != .unavailable, let value = status.coverage, !value.isNaN else {
    return nil
  }

  let coverage = min(1, max(0, value))
  let hue = 120 * (1 - sqrt(1 - coverage))
  let saturation = 0.75
  let lightness = 0.55
  let chroma = (1 - abs(2 * lightness - 1)) * saturation
  let secondary = chroma * (1 - abs((hue / 60).truncatingRemainder(dividingBy: 2) - 1))
  let offset = lightness - chroma / 2
  let (red, green) = hue < 60 ? (chroma, secondary) : (secondary, chroma)
  return PaceColor(
    red: Int(((red + offset) * 255).rounded()),
    green: Int(((green + offset) * 255).rounded()),
    blue: Int((offset * 255).rounded())
  )
}

func displayPercent(_ value: Double) -> Double {
  min(100, max(0, value))
}

func normalizedPercent(_ value: Double) -> Double {
  (displayPercent(value) * 10).rounded() / 10
}

func severity(_ value: Double) -> QuotaLevel {
  if value >= 90 { return .critical }
  if value >= 70 { return .warning }
  return .normal
}

func formatPercent(_ value: Double?) -> String {
  guard let value else { return "—" }
  let displayed = displayPercent(value)
  return displayed.rounded() == displayed
    ? String(format: "%.0f%%", displayed)
    : String(format: "%.1f%%", displayed)
}

func formatTime(_ value: Date?) -> String {
  value?.formatted(date: .omitted, time: .shortened) ?? "—"
}

func formatReset(_ value: Date?) -> String {
  value.map { "resets \($0.formatted(date: .numeric, time: .shortened))" }
    ?? "reset unknown"
}

File: windows/CuotaX/Quota.cs
// SPDX-License-Identifier: MIT

using System.Globalization;

namespace CuotaX;

internal sealed record QuotaWindow(double UsedPercent, DateTimeOffset? ResetsAt);

internal sealed record Quota(QuotaWindow? FiveHour, QuotaWindow? Weekly, DateTimeOffset UpdatedAt)
{
    internal double? HighestPercent
    {
        get
        {
            var values = new[] { FiveHour, Weekly }
                .Where(value => value is not null)
                .Select(value => value!.UsedPercent)
                .ToArray();
            return values.Length == 0 ? null : values.Max();
        }
    }

    internal static string? ResetMessage(Quota? previous, Quota current)
    {
        if (previous is null)
        {
            return null;
        }

        var fiveHour = CrossedReset(previous.FiveHour, previous.UpdatedAt, current.UpdatedAt);
        var weekly = CrossedReset(previous.Weekly, previous.UpdatedAt, current.UpdatedAt);
        return (fiveHour, weekly) switch
        {
            (true, true) => "5-hour and weekly quotas reset",
            (true, false) => "5-hour quota reset",
            (false, true) => "Weekly quota reset",
            _ => null,
        };
    }

    internal DateTimeOffset? NextReset(DateTimeOffset? now = null)
    {
        var current = now ?? DateTimeOffset.Now;
        return new[] { FiveHour?.ResetsAt, Weekly?.ResetsAt }
            .Where(value => value > current)
            .Min();
    }

    internal QuotaStatus Status(DateTimeOffset? now = null)
    {
        var current = now ?? DateTimeOffset.Now;
        var statuses = new[]
        {
            WindowStatus(FiveHour, "5-hour", QuotaFormatting.FiveHourMinutes, current),
            WindowStatus(Weekly, "weekly", QuotaFormatting.WeeklyMinutes, current),
        }.Where(value => value is not null).Select(value => value!).ToArray();

        if (statuses.Length == 0)
        {
            return QuotaStatus.Unavailable;
        }

        var timed = statuses.Where(value => value.Status.OnTrackPercent is not null).ToArray();
        return (timed.Length == 0 ? statuses : timed).MinBy(value => value.Coverage)!.Status;
    }

    internal static QuotaStatus StatusFor(
        QuotaWindow? value,
        string window,
        int durationMinutes,
        DateTimeOffset? now = null
    ) => WindowStatus(value, window, durationMinutes, now ?? DateTimeOffset.Now)?.Status
        ?? QuotaStatus.Unavailable;

    private static RankedStatus? WindowStatus(
        QuotaWindow? value,
        string window,
        int durationMinutes,
        DateTimeOffset now
    )
    {
        if (value is null)
        {
            return null;
        }

        var usedPercent = QuotaFormatting.NormalizePercent(value.UsedPercent);
        var remainingPercent = 100 - usedPercent;
        if (value.ResetsAt is null || value.ResetsAt <= now)
        {
            var untimedCoverage = remainingPercent / 100;
            return new RankedStatus(
                untimedCoverage,
                new QuotaStatus(
                    untimedCoverage,
                    QuotaFormatting.Severity(usedPercent),
                    null,
                    null,
                    remainingPercent,
                    window
                )
            );
        }

        var duration = TimeSpan.FromMinutes(durationMinutes).TotalSeconds;
        var targetEnd = durationMinutes == QuotaFormatting.WeeklyMinutes
            ? new[] { NextLocalMidnight(now), value.ResetsAt.Value }.Min()
            : now;
        var targetTimeRemaining = Math.Clamp(
            (value.ResetsAt.Value - targetEnd).TotalSeconds / duration,
            0,
            1
        );
        var actualTimeRemaining = Math.Clamp(
            (value.ResetsAt.Value - now).TotalSeconds / duration,
            0,
            1
        );
        var targetPercent = 100 * (1 - targetTimeRemaining);
        var onTrackPercent = Math.Round(
            10 * targetPercent,
            MidpointRounding.AwayFromZero
        ) / 10;
        double? paceProgress = null;
        if (durationMinutes == QuotaFormatting.WeeklyMinutes)
        {
            var startTimeRemaining = Math.Clamp(
                (value.ResetsAt.Value - StartOfLocalDay(now)).TotalSeconds / duration,
                0,
                1
            );
            var startPercent = 100 * (1 - startTimeRemaining);
            var allowance = targetPercent - startPercent;
            if (allowance > 0)
            {
                paceProgress = (usedPercent - startPercent) / allowance;
            }
        }
        var coverage = actualTimeRemaining == 0
            ? double.PositiveInfinity
            : remainingPercent / 100 / actualTimeRemaining;
        var level = usedPercent <= onTrackPercent
            ? QuotaLevel.Normal
            : coverage < 0.5
                ? QuotaLevel.Critical
                : QuotaLevel.Warning;

        return new RankedStatus(
            coverage,
            new QuotaStatus(coverage, level, onTrackPercent, paceProgress, remainingPercent, window)
        );
    }

    private static bool CrossedReset(
        QuotaWindow? window,
        DateTimeOffset previousUpdatedAt,
        DateTimeOffset currentUpdatedAt
    ) => window?.ResetsAt is { } resetsAt
        && previousUpdatedAt < resetsAt
        && resetsAt <= currentUpdatedAt;

    private static DateTimeOffset NextLocalMidnight(DateTimeOffset now)
        => LocalMidnight(now, 1);

    private static DateTimeOffset StartOfLocalDay(DateTimeOffset now)
        => LocalMidnight(now, 0);

    private static DateTimeOffset LocalMidnight(DateTimeOffset now, int days)
    {
        var local = TimeZoneInfo.ConvertTime(now, TimeZoneInfo.Local);
        var midnight = DateTime.SpecifyKind(local.Date.AddDays(days), DateTimeKind.Unspecified);
        return new DateTimeOffset(midnight, TimeZoneInfo.Local.GetUtcOffset(midnight));
    }

    private sealed record RankedStatus(double Coverage, QuotaStatus Status);
}

internal enum QuotaLevel
{
    Normal,
    Warning,
    Critical,
    Unavailable,
}

internal sealed record QuotaStatus(
    double? Coverage,
    QuotaLevel Level,
    double? OnTrackPercent,
    double? PaceProgress,
    double? RemainingPercent,
    string? Window
)
{
    internal static QuotaStatus Unavailable { get; } = new(
        null,
        QuotaLevel.Unavailable,
        null,
        null,
        null,
        null
    );

    internal string PaceLabel => Level switch
    {
        QuotaLevel.Normal => "On track",
        QuotaLevel.Warning => "Running high",
        QuotaLevel.Critical => "At risk",
        _ => "Pace unavailable",
    };
}

internal readonly record struct PaceColor(int Red, int Green, int Blue);

internal static class QuotaFormatting
{
    internal const int FiveHourMinutes = 5 * 60;
    internal const int WeeklyMinutes = 7 * 24 * 60;

    internal static double DisplayPercent(double value) => Math.Clamp(value, 0, 100);

    internal static double NormalizePercent(double value) => Math.Round(
        DisplayPercent(value) * 10,
        MidpointRounding.AwayFromZero
    ) / 10;

    internal static QuotaLevel Severity(double value) => value switch
    {
        >= 90 => QuotaLevel.Critical,
        >= 70 => QuotaLevel.Warning,
        _ => QuotaLevel.Normal,
    };

    internal static PaceColor? Color(QuotaStatus status)
    {
        if (status.Level == QuotaLevel.Unavailable || status.RemainingPercent is null || !double.IsFinite(status.RemainingPercent.Value))
        {
            return null;
        }

        var usage = 1 - Math.Clamp(status.RemainingPercent.Value, 0, 100) / 100;
        var threshold = Math.Clamp((status.OnTrackPercent ?? 70) / 100, 0, 1);
        double hue;
        if (status.PaceProgress is { } progress && double.IsFinite(progress) && usage <= threshold)
        {
            hue = 120 - 90 * Smoothstep(Math.Clamp(progress, 0, 1));
        }
        else if (usage <= 0)
        {
            hue = 120;
        }
        else if (usage < threshold)
        {
            hue = 120 - 90 * Smoothstep(usage / threshold);
        }
        else if (usage < 1)
        {
            hue = 30 * (1 - Smoothstep((usage - threshold) / (1 - threshold)));
        }
        else
        {
            hue = 0;
        }
        const double saturation = 0.75;
        const double lightness = 0.55;
        var chroma = (1 - Math.Abs(2 * lightness - 1)) * saturation;
        var secondary = chroma * (1 - Math.Abs((hue / 60 % 2) - 1));
        var offset = lightness - chroma / 2;
        var (red, green) = hue < 60 ? (chroma, secondary) : (secondary, chroma);
        return new PaceColor(
            (int)Math.Round((red + offset) * 255, MidpointRounding.AwayFromZero),
            (int)Math.Round((green + offset) * 255, MidpointRounding.AwayFromZero),
            (int)Math.Round(offset * 255, MidpointRounding.AwayFromZero)
        );
    }

    private static double Smoothstep(double value) => value * value * (3 - 2 * value);

    internal static string Percent(double? value)
    {
        if (value is null)
        {
            return "—";
        }

        var displayed = DisplayPercent(value.Value);
        var format = displayed == Math.Round(displayed) ? "0" : "0.0";
        return $"{displayed.ToString(format, CultureInfo.CurrentCulture)}%";
    }

    internal static string Time(DateTimeOffset? value) => value?.ToLocalTime().ToString("t", CultureInfo.CurrentCulture) ?? "—";

    internal static string Reset(DateTimeOffset? value, DateTimeOffset? now = null)
    {
        if (value is null)
        {
            return "reset unknown";
        }

        var remaining = Remaining(value.Value - (now ?? DateTimeOffset.Now));
        var suffix = remaining is null ? "" : $" ({remaining})";
        return $"{value.Value.ToLocalTime().ToString("ddd, MMM d HH:mm", CultureInfo.InvariantCulture)}{suffix}";
    }

    private static string? Remaining(TimeSpan value)
    {
        if (value <= TimeSpan.Zero)
        {
            return null;
        }
        if (value.TotalDays >= 1)
        {
            return $"{(int)value.TotalDays}d";
        }
        if (value.TotalHours >= 1)
        {
            return $"{(int)value.TotalHours}h";
        }
        return $"{Math.Max(1, (int)value.TotalMinutes)}m";
    }
}
